src/Integration.Vsix/Delegates/SonarAnalyzerConnectedWorkflow.cs

```diff
@@ -0,0 +1,165 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+using SonarLint.VisualStudio.Integration.Persistence;
+using SonarLint.VisualStudio.Integration.Rules;
+using SonarLint.VisualStudio.Integration.Vsix.Suppression;
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    internal class SonarAnalyzerConnectedWorkflow : SonarAnalyzerWorkflowBase
+    {
+        private readonly IQualityProfileProvider qualityProfileProvider;
+        private readonly BoundSonarQubeProject boundProject;
+        private readonly IDictionary<Language, QualityProfile> cachedQualityProfiles;
+        private readonly ISuppressionHandler suppressionHandler;
+        private readonly Func<IAnalysisRunContext, bool> previousShouldExecuteRuleFunc;
+        private readonly Action<IReportingContext> previousReportDiagnosticAction;
+
+        public SonarAnalyzerConnectedWorkflow(Workspace workspace, IQualityProfileProvider qualityProfileProvider,
+            BoundSonarQubeProject boundProject, ISuppressionHandler suppressionHandler)
+            : base(workspace)
+        {
+            if (qualityProfileProvider == null)
+            {
+                throw new ArgumentNullException(nameof(qualityProfileProvider));
+            }
+
+            if (boundProject == null)
+            {
+                throw new ArgumentNullException(nameof(boundProject));
+            }
+
+            if (suppressionHandler == null)
+            {
+                throw new ArgumentNullException(nameof(suppressionHandler));
+            }
+
+            this.qualityProfileProvider = qualityProfileProvider;
+            this.boundProject = boundProject;
+            this.suppressionHandler = suppressionHandler;
+            this.cachedQualityProfiles = FetchSupportedQualityProfiles();
+
+            // Saving previous state so that SonarLint doesn't have to know what's the default state in SonarAnalyzer
```

Comment:
Wouldn't the default state always be the same as Standalone?

Reply:
On my first test I was not saving the previous state and I was resetting the delegates to null during the `Dispose()`. Then I thought that changing the initial state in the analyzer would break this code so I introduced this save-and-restore on the different workflows. If you think the first solution was good enough I don't mind getting back to it.

Reply:
As discussed offline, moving the save-and-reset could be moved up to the manager would solve most of the issues.